### PR TITLE
Add iptables firewall support to openshift-common

### PR DESCRIPTION
**Pull request by detiber**
- Set base_os to default to not installing firewalld when the host has a role
  of openshift_master or openshift_node
- Hardcoded openshift_common to use iptables (through the vars directory)
  until upstream support is in place for firewalld
- Added a iptables firewall module for maintaining the following (in a mostly
  naive manner):
  - ensure the OPENSHIFT_ALLOW chain is defined
  - ensure that there is a jump rule in the INPUT chain for OPENSHIFT_ALLOW
  - adds or removes entries from the OPENSHIFT_ALLOW chain
  - issues '/usr/libexec/iptables/iptables.init save' when rules are changed
- Limitations of iptables firewall module
  - only allows setting of ports/protocols to open
  - no testing on ipv6 support

### CODE REVIEW

**Comment by detiber** [comment]
@twiest @jwhonce : the iptables related change I mentioned yesterday.

File: roles/base_os/defaults/main.yml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+---
+base_os_use_firewalld: "{{ False if ('openshift_master' in role_names or 'openshift_node' in role_names) else True }}"
```
Comment (twiest) [inline roles/base_os/defaults/main.yml]: We prefer to keep all roles as generic as possible. This makes the base_os role 'know' about the openshift_master and openshift_node roles. Also, if we want to add more exclusions in the future, we'd have to add more to this list, which gets ugly quickly.

Instead, consider just passing in this variable to this role in the specific config playbooks for the openshift_master and openshift_node host types.

File: roles/openshift_common/library/openshift_manage_iptables.py [hunk 2]
```diff
@@ -0,0 +1,254 @@
+#!/usr/bin/python
```
Comment (twiest) [inline roles/openshift_common/library/openshift_manage_iptables.py]: To keep our python code as consistent as possible, we've been adding the following line to the top of all of our python scripts. I know it doesn't solve all consistency problems, but it solves a lot of them.

# vim: expandtab:tabstop=4:shiftwidth=4

**Comment by twiest** [comment]
@detiber I like where you're going here.

I was thinking, though, maybe the firewall stuff should be in it's own 'os_firewall' role.

Then we could remove all firewall setup code from base_os and openshift_common and have have other rolls / playbooks call into os_firewall directly. I'd really like to share this code for multiple host types and roles. The os_firewall role would, of course, abstract away the iptables / firewalld things that you've done here in openshift_common. Thoughts?

File: roles/base_os/defaults/main.yml (hunk 1 above)
Comment (detiber) [inline roles/base_os/defaults/main.yml]: @twiest The easiest way would probably to remove the dependency on base_os (which is really at this point just making sure the firewall is turned on, installing bash_completion and setting the .vimrc and .irbrc for the root user. I would rather see a better way to apply those things on dev hosts without bringing them in for origin/enterprise installs. Any objections to moving those to a new role that is conditionally applied to hosts...  something like os_env_extras or something?

That said, my main idea with setting the default there is that I wasn't sure if the base_os role would have access to the variables that are set in the openshift_master and openshift_node roles, since the base_os role is processed before they are.

**Comment by detiber** [comment]
@twiest +1 on abstracting into a role. I'll work through the change.  I would like to add a dependency on os_firewall (I like the os_ convention for dependency classes that aren't openshift_ only) to either the openshift_common or both openshift_{master,node} so that we pull it in automatically without extra configuration.  I can see users wanting to write their own playbooks around managing openshift.

File: roles/openshift_common/library/openshift_manage_iptables.py (hunk 2 above)
Comment (detiber) [inline roles/openshift_common/library/openshift_manage_iptables.py]: I went ahead and added the modeline to both the openshift_common/library/openshift_manage_iptables.py module and the openshift_node/library/openshift_register_node.py module.

File: roles/os_firewall/vars/main.yml [hunk 3]
```diff
@@ -0,0 +1,4 @@
+---
+os_firewall_pkgs: "{{ ['firewalld'] if os_firewall_use_firewalld else ['iptables', 'iptables-services'] }}"
+os_firewall_svc: "{{ 'firewalld' if os_firewall_use_firewalld else 'iptables' }}"
+os_firewall_mask_svcs: "{{ ['iptables'] if os_firewall_use_firewalld else ['firewalld'] }}"
```
Comment (twiest) [inline roles/os_firewall/vars/main.yml]: So I see that here you're trying to abstract out the logic so that we can have 1 yum command and it'll install either iptables or firewalld packages.

While I commend you for trying to keep with the DRY principle (which I subscribe too as well), what I prefer in this situation is that firewall/firewalld.yml have the yum calls for firewalld specifically, and that firewall/iptables.yml have the yum calls for iptables.

Same for the services, masking, etc.

It makes the logic much easier to follow, and any errors will be much clearer. Otherwise I have to jump through multiple files to follow the logic of how stuff got set to each value.

I don't believe this violates DRY, as I see these lines the same as method calls, whereas DRY is specifically for not defining multiple methods with the same business logic.

Also, I prefer to not do too much logic in jinja as the errors aren't as clearly laid out as they are in ansible, and they're harder to debug.

In other words, I prefer stuff like this (very simple to follow and debug):
```
# start of firewall/firewalld.yml
- name: Install firewalld packages
  yum: name={{ item }} state=present
  with_items:
    - firewalld

# start of firewall/iptables.yml
- name: Install iptables packages
  yum: name={{ item }} state=present
  with_items:
    - iptables
    - iptables-services
```

**Comment by twiest** [comment]
@detiber I like where this is going, let me know when it's ready for another review / merge.

**Comment by detiber** [comment]
@twiest latest firewall changes, also moving the renaming the base_os rule to os_env_extras.

If these changes look good, then I would like to squash before this gets committed.

File: roles/openshift_common/vars/main.yml [hunk 4]
```diff
@@ -1,2 +1,9 @@
 ---
 openshift_master_credentials_dir: /var/lib/openshift/openshift.local.certificates/admin/
+
+# TODO: Upstream kubernetes only supports iptables currently, if this changes,
+# then these variable should be moved to defaults
+openshift_use_firewalld: False
+openshift_fw_pkgs: "{{ ['firewalld'] if openshift_use_firewalld else ['iptables', 'iptables-services'] }}"
+openshift_fw_svc: "{{ 'firewalld' if openshift_use_firewalld else 'iptables' }}"
+openshift_fw_mask_svcs: "{{ ['iptables'] if openshift_use_firewalld else ['firewalld'] }}"
```
Comment (twiest) [inline roles/openshift_common/vars/main.yml]: With your other changes, these 3 variables are no longer used AFAICS.

**Comment by twiest** [comment]
@detiber this is looking excellent!  Just one minor cleanup comment above, but other than that, I think it's ready to merge.

**Comment by detiber** [comment]
squashed and rebased with the fix for removing the extraneous variables.

**Comment by twiest** [comment]
@detiber Looks great!  :+1:
